### DISCUSSION TITLE
Added a global scope to User model, role method always returns empty collection

### DISCUSSION THREAD

**Discussion by cjaoude** [spatie/laravel-permission#1731] [Q&A] [answered]
I extended the User class for a role. (Example) I gave Student a global scope for the 'student' role. This way, rather than call `User::student()->fooQuery()->get()` all over the place, I can do `Student::fooQuery()->get()`. Plus other uses of course.

The global scope works just fine with Eloquent conditions (ie. where clause, etc), but the second I use the spatie role methods i always get back an empty collection.

Any ideas why this is not working? Tried it with both `boot()` and `booted()`, same result.

```php
class Student extends User
{
    /**
     * The "booted" method of the model.
     *
     * @return void
     */
    protected static function booted()
    {
        parent::booted();

        static::addGlobalScope('foobar', function (Builder $builder) {
            $builder->role('student'); // always an empty collection
        });
    }
}
```

**Comment by drbyte** [comment, marked as answer]
Given that I don't currently have an app set up that would let me step through the code to figure out what might be contributing to the matter, I offer the following which is where I'd start exploring in such an app:

- I'd check whether the `scopeRole()` method on the trait is actually being called, or if something else is being triggered. Maybe add a `dd()` in there?
- in your closure maybe `dd()` the `$builder` to check whether there's a mismatch of objects used vs expected
- consider cloning a bunch of the [`scopeRole()`](https://github.com/spatie/laravel-permission/blob/f0fe2c879478f332b72633b14468c98a7f2c84c5/src/Traits/HasRoles.php#L60-L84) method into your `addGlobalScope()`, but using only as little as possible. Perhaps start with just its return closure, after reducing the `$roles` variable value to just the object it represents after all the code there is done, so that the return is simple. 
Cloning that into your `addGlobalScope` and then paring it down may reveal the problem/issue behind your empty collection result, and then perhaps your needed solution will become self evident.

Hope that helps!

**Reply by cjaoude** [reply]
I figured it out. Need to tell the global scope to use the parent model

```
        static::addGlobalScope('foo', function (Builder $builder) {
            $builder->setModel(new User)->role(User::FOO_ROLE);
        });
```

**Reply by cjaoude** [reply]
Actually, further figured out this was necessary because the pivot table for assigned roles was using User, rather than the class extending User. So it comes down to pros & cons of this approach. But at least now get what's happening.
